src/software/ai/evaluation/find_passes.cpp

```diff
@@ -0,0 +1,101 @@
+                                   const Team& enemy_team)
+{
+    std::vector<Robot> open_robots =
+        findOpenFriendlyRobots(friendly_team, enemy_team, ROBOT_MAX_RADIUS_METERS * 1.5);
+
+    open_robots.erase(remove(open_robots.begin(), open_robots.end(), robot));
+    std::vector<Robot> robot_obstacles = friendly_team.getAllRobots();
+    robot_obstacles.insert(robot_obstacles.begin(), enemy_team.getAllRobots().begin(),
+                           enemy_team.getAllRobots().end());
+    robot_obstacles.erase(remove(robot_obstacles.begin(), robot_obstacles.end(), robot));
+
+    std::vector<Circle> obstacles =
+        getObstacles(robot_obstacles, ROBOT_MAX_RADIUS_METERS * 1.5);
+    std::vector<Pass> direct_passes;
+    for (Robot open_robot : open_robots)
+    {
+        Circle obstacle = getObstacle(open_robot, ROBOT_MAX_RADIUS_METERS * 1.5);
+        obstacles.erase(remove(obstacles.begin(), obstacles.end(), obstacle));
+        Segment possible_pass = Segment(robot.position(), open_robot.position());
+        if (std::all_of(obstacles.begin(), obstacles.end(),
+                        [possible_pass](Circle obstacle) {
+                            return !intersects(possible_pass, obstacle);
+                        }))
+        {
+            direct_passes.push_back(Pass(robot.position(), open_robot.position(),
+                                         BALL_MAX_SPEED_METERS_PER_SECOND));
+        }
+        obstacles.push_back(obstacle);
+    }
+    return direct_passes;
+}
+
+std::vector<Pass> findIndirectPasses(const Robot& robot, const Team& friendly_team,
+                                     const Team& enemy_team)
+{
+    std::vector<Robot> open_robots =
+        findOpenFriendlyRobots(friendly_team, enemy_team, ROBOT_MAX_RADIUS_METERS * 1.5);
+    open_robots.erase(remove(open_robots.begin(), open_robots.end(), robot));
+
+    std::vector<Robot> robot_obstacles = friendly_team.getAllRobots();
+    robot_obstacles.insert(robot_obstacles.begin(), enemy_team.getAllRobots().begin(),
+                           enemy_team.getAllRobots().end());
+    robot_obstacles.erase(remove(robot_obstacles.begin(), robot_obstacles.end(), robot));
+
+    std::vector<Circle> obstacles =
+        getObstacles(robot_obstacles, ROBOT_MAX_RADIUS_METERS * 1.5);
+    std::vector<Pass> indirect_passes;
+    for (Robot open_robot : open_robots)
+    {
+        Circle obstacle = getObstacle(open_robot, ROBOT_MAX_RADIUS_METERS * 1.5);
+        obstacles.erase(remove(obstacles.begin(), obstacles.end(), obstacle));
+        Segment possible_pass = Segment(robot.position(), open_robot.position());
+        if (std::any_of(obstacles.begin(), obstacles.end(),
+                        [possible_pass](Circle obstacle) {
+                            return intersects(possible_pass, obstacle);
+                        }))
+        {
+            indirect_passes.push_back(Pass(robot.position(), open_robot.position(),
```

Comment:
have you considered only returning the end positions of open passes instead of a full pass (vector<Point>)?

From what I see, each pass is only different by one argument and it seems like we'll only pick one pass out of the list we get at each tick, so it seems pointless to construct all these passes only to throw them away later

Reply:
That's a good point. I changed it so it returns two vectors of points instead.